File: tests/src/org.joda/joda-convert/2.2.4/src/test/java/org_joda/joda_convert/StringConvertTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_joda.joda_convert;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;

import java.io.File;
import java.lang.module.Configuration;
import java.lang.module.FindException;
import java.lang.module.ModuleFinder;
import java.lang.reflect.Type;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.OptionalDouble;
import java.util.OptionalInt;
import java.util.OptionalLong;
import java.util.Set;

import org.graalvm.internal.tck.NativeImageSupport;
import org.joda.convert.RenameHandler;
import org.joda.convert.StringConvert;
import org.junit.jupiter.api.Test;

public class StringConvertTest {
    @Test
    public void registersMethodsAndConstructorsByName() {
        StringConvert convert = new StringConvert();
        convert.registerMethods(MethodValue.class, "asText", "parse");
        convert.registerMethods(CharSequenceMethodValue.class, "asText", "parse");
        convert.registerMethodConstructor(File.class, "getPath");
        convert.registerMethodConstructor(CharSequenceConstructorFixture.class, "asText");

        assertEquals("alpha", convert.convertToString(new MethodValue("alpha")));
        assertEquals("beta", convert.convertFromString(MethodValue.class, "beta").asText());
        assertEquals("gamma", convert.convertFromString(CharSequenceMethodValue.class, "gamma").asText());
        assertEquals("build.gradle", convert.convertFromString(File.class, "build.gradle").getPath());
        assertEquals(
                "delta",
                convert.convertFromString(CharSequenceConstructorFixture.class, "delta").asText());
    }

    @Test
    public void convertsJava8OptionalTypesRegisteredDuringConstruction() {
        StringConvert convert = new StringConvert();

        assertEquals("12", convert.convertToString(OptionalInt.of(12)));
        assertEquals("", convert.convertToString(OptionalInt.empty()));
        assertEquals(OptionalInt.of(34), convert.convertFromString(OptionalInt.class, "34"));
        assertEquals(OptionalInt.empty(), convert.convertFromString(OptionalInt.class, ""));

        assertEquals("56", convert.convertToString(OptionalLong.of(56L)));
        assertEquals(OptionalLong.of(78L), convert.convertFromString(OptionalLong.class, "78"));

        assertEquals("1.25", convert.convertToString(OptionalDouble.of(1.25d)));
        assertEquals(OptionalDouble.of(2.5d), convert.convertFromString(OptionalDouble.class, "2.5"));
    }

    @Test
    public void loadsTypesWithoutAContextClassLoader() throws Exception {
        Thread thread = Thread.currentThread();
        ClassLoader originalLoader = thread.getContextClassLoader();
        try {
            thread.setContextClassLoader(null);
            assertSame(String.class, RenameHandler.create(false).lookupType("java.lang.String"));
        } finally {
            thread.setContextClassLoader(originalLoader);
        }
    }

    @Test
    public void registersGuavaConverterWhenLoadedAsNamedModules() throws Exception {
        try {
            assertEquals("java.lang.String", convertGuavaTypeTokenInNamedModuleLayer());
        } catch (Error error) {
            if (NativeImageSupport.isUnsupportedFeatureError(error)) {
                return;
            }
            throw error;
        } catch (FindException exception) {
            // Native Image code sources point at the executable, not the original modular JAR.
            return;
        }
    }

    private static String convertGuavaTypeTokenInNamedModuleLayer() throws Exception {
        Path jodaConvertJar = jarPath(StringConvert.class);
        Class<?> classPathTypeTokenClass = Class.forName("com.google.common.reflect.TypeToken");
        Path guavaJar = jarPath(classPathTypeTokenClass);
        ModuleFinder finder = ModuleFinder.of(jodaConvertJar, guavaJar);
        String jodaConvertModuleName = moduleName(finder, "org.joda.convert");
        String guavaModuleName = moduleName(finder, "com.google.common.reflect");
        Configuration configuration = ModuleLayer.boot().configuration()
                .resolve(finder, ModuleFinder.of(), Set.of(jodaConvertModuleName, guavaModuleName));
        ModuleLayer layer = ModuleLayer.boot()
                .defineModulesWithOneLoader(configuration, ClassLoader.getPlatformClassLoader());
        ClassLoader loader = layer.findLoader(jodaConvertModuleName);
        Thread thread = Thread.currentThread();
        ClassLoader originalLoader = thread.getContextClassLoader();
        try {
            thread.setContextClassLoader(loader);
            Class<?> stringConvertType = Class.forName("org.joda.convert.StringConvert", true, loader);
            Object convert = stringConvertType.getConstructor().newInstance();
            Class<?> typeTokenClass = Class.forName("com.google.common.reflect.TypeToken", true, loader);
            Object token = typeTokenClass.getMethod("of", Type.class).invoke(null, String.class);
            return (String) stringConvertType.getMethod("convertToString", Class.class, Object.class)
                    .invoke(convert, typeTokenClass, token);
        } finally {
            thread.setContextClassLoader(originalLoader);
        }
    }

    private static Path jarPath(Class<?> type) throws URISyntaxException {
        return Paths.get(type.getProtectionDomain().getCodeSource().getLocation().toURI());
    }

    private static String moduleName(ModuleFinder finder, String packageName) {
        return finder.findAll().stream()
                .filter(reference -> reference.descriptor().packages().contains(packageName))
                .findFirst()
                .orElseThrow(() -> new IllegalStateException("No module contains package " + packageName))
                .descriptor()
                .name();
    }

    public static final class MethodValue {
        private final String value;

        MethodValue(String value) {
            this.value = value;
        }

        public String asText() {
            return value;
        }

        public static MethodValue parse(String value) {
            return new MethodValue(value);
        }
    }

    public static final class CharSequenceMethodValue {
        private final String value;

        CharSequenceMethodValue(String value) {
            this.value = value;
        }

        public String asText() {
            return value;
        }

        public static CharSequenceMethodValue parse(CharSequence value) {
            return new CharSequenceMethodValue(value.toString());
        }
    }
}

File: tests/src/org.playframework.anorm/anorm_3/2.8.1/src/test/scala/org_playframework_anorm/anorm_3/StringConvertTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_playframework_anorm.anorm_3

import java.lang.reflect.Type
import java.util.OptionalDouble
import java.util.OptionalInt
import java.util.OptionalLong

import com.google.common.reflect.TypeToken
import org.joda.convert.StringConvert
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

import scala.annotation.static

class StringConvertTest {
  @Test
  def constructorRegistersGuavaAndJavaOptionalConverters(): Unit = {
    val convert: StringConvert = new StringConvert()

    val optionalInt: OptionalInt = convert.convertFromString(classOf[OptionalInt], "17")
    val optionalLong: OptionalLong = convert.convertFromString(classOf[OptionalLong], "19")
    val optionalDouble: OptionalDouble = convert.convertFromString(classOf[OptionalDouble], "23.5")
    val token: TypeToken[?] = convert.convertFromString(
      classOf[TypeToken[?]],
      "java.util.List<java.lang.String>"
    )

    assertEquals(17, optionalInt.getAsInt)
    assertEquals(19L, optionalLong.getAsLong)
    assertEquals(23.5d, optionalDouble.getAsDouble, 0.0d)
    assertEquals("17", convert.convertToString(optionalInt))
    assertEquals("19", convert.convertToString(optionalLong))
    assertEquals("23.5", convert.convertToString(optionalDouble))
    assertEquals("java.util.List<java.lang.String>", token.toString)
  }

  @Test
  def typeParsingUsesTheContextClassLoaderWhenItIsAvailable(): Unit = {
    val convert: StringConvert = new StringConvert()

    val parsed: Type = convert.convertFromString(
      classOf[Type],
      "java.util.Map<java.lang.String, java.lang.Integer>"
    )

    assertEquals("java.util.Map<java.lang.String, java.lang.Integer>", parsed.getTypeName)
  }

  @Test
  def typeParsingUsesClassForNameWhenThereIsNoContextClassLoader(): Unit = {
    val thread: Thread = Thread.currentThread()
    val originalLoader: ClassLoader = thread.getContextClassLoader
    thread.setContextClassLoader(null)
    try {
      val convert: StringConvert = new StringConvert()
      val parsed: Type = convert.convertFromString(classOf[Type], "java.lang.String")

      assertEquals(classOf[String], parsed)
    } finally {
      thread.setContextClassLoader(originalLoader)
    }
  }

  @Test
  def registerMethodsAcceptsStaticStringFactoryMethods(): Unit = {
    val convert: StringConvert = new StringConvert(false)
    convert.registerMethods(classOf[StringMethodValue], "asText", "parse")

    val parsed: StringMethodValue = convert.convertFromString(classOf[StringMethodValue], "from-string-method")

    assertEquals("from-string-method", parsed.asText)
    assertEquals("from-string-method", convert.convertToString(parsed))
  }

  @Test
  def registerMethodsFallsBackToStaticCharSequenceFactoryMethods(): Unit = {
    val convert: StringConvert = new StringConvert(false)
    convert.registerMethods(classOf[CharSequenceMethodValue], "asText", "parse")

    val parsed: CharSequenceMethodValue = convert.convertFromString(
      classOf[CharSequenceMethodValue],
      "from-char-sequence-method"
    )

    assertEquals("from-char-sequence-method", parsed.asText)
    assertEquals("from-char-sequence-method", convert.convertToString(parsed))
  }

  @Test
  def registerMethodConstructorAcceptsStringConstructors(): Unit = {
    val convert: StringConvert = new StringConvert(false)
    convert.registerMethodConstructor(classOf[StringConstructorValue], "asText")

    val parsed: StringConstructorValue = convert.convertFromString(
      classOf[StringConstructorValue],
      "from-string-constructor"
    )

    assertEquals("from-string-constructor", parsed.asText)
    assertEquals("from-string-constructor", convert.convertToString(parsed))
  }

  @Test
  def registerMethodConstructorFallsBackToCharSequenceConstructors(): Unit = {
    val convert: StringConvert = new StringConvert(false)
    convert.registerMethodConstructor(classOf[CharSequenceConstructorValue], "asText")

    val parsed: CharSequenceConstructorValue = convert.convertFromString(
      classOf[CharSequenceConstructorValue],
      "from-char-sequence-constructor"
    )

    assertEquals("from-char-sequence-constructor", parsed.asText)
    assertEquals("from-char-sequence-constructor", convert.convertToString(parsed))
  }
}

final class StringMethodValue(private val value: String) {
  def asText: String = value
}

object StringMethodValue {
  @static def parse(value: String): StringMethodValue = new StringMethodValue(value)
}

final class CharSequenceMethodValue(private val value: CharSequence) {
  def asText: String = value.toString
}

object CharSequenceMethodValue {
  @static def parse(value: CharSequence): CharSequenceMethodValue = new CharSequenceMethodValue(value)
}

final class StringConstructorValue(private val value: String) {
  def asText: String = value
}

final class CharSequenceConstructorValue(private val value: CharSequence) {
  def asText: String = value.toString
}
